src/contrib/cluster/Akka.Cluster.Sharding/DDataShard.cs

```diff
@@ -0,0 +1,229 @@
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.DistributedData;
+using Akka.Event;
+
+namespace Akka.Cluster.Sharding
+{
+    using ShardId = String;
+    using EntryId = String;
+    using Msg = Object;
+
+    /// <summary>
+    /// This actor creates children entity actors on demand that it is told to be
+    /// responsible for. It is used when `rememberEntities` is enabled and
+    /// `state-store-mode=ddata`.
+    /// </summary>
+    internal sealed class DDataShard : ActorBase, IShard, IWithUnboundedStash
+    {
+        IActorContext IShard.Context => Context;
+        IActorRef IShard.Self => Self;
+        IActorRef IShard.Sender => Sender;
+        void IShard.Unhandled(object message) => base.Unhandled(message);
+
+        public string TypeName { get; }
+        public string ShardId { get; }
+        public Props EntityProps { get; }
+        public ClusterShardingSettings Settings { get; }
+        public ExtractEntityId ExtractEntityId { get; }
+        public ExtractShardId ExtractShardId { get; }
+        public object HandOffStopMessage { get; }
+        ILoggingAdapter IShard.Log { get; } = Context.GetLogger();
+        public IActorRef HandOffStopper { get; set; }
+        public Shard.ShardState State { get; set; } = Shard.ShardState.Empty;
+        public ImmutableDictionary<string, IActorRef> RefById { get; set; } = ImmutableDictionary<string, IActorRef>.Empty;
+        public ImmutableDictionary<IActorRef, string> IdByRef { get; set; } = ImmutableDictionary<IActorRef, string>.Empty;
+        public ImmutableHashSet<IActorRef> Passivating { get; set; } = ImmutableHashSet<IActorRef>.Empty;
+        public ImmutableDictionary<string, ImmutableList<Tuple<object, IActorRef>>> MessageBuffers { get; set; } = ImmutableDictionary<string, ImmutableList<Tuple<object, IActorRef>>>.Empty;
+
+        private EntityRecoveryStrategy RememberedEntitiesRecoveryStrategy { get; }
+        public Cluster Cluster { get; } = Cluster.Get(Context.System);
+        public ILoggingAdapter Log { get; } = Context.GetLogger();
+        public IActorRef Replicator { get; }
+        public int MajorityCap { get; }
+        public IStash Stash { get; set; }
+
+        private readonly IReadConsistency _readConsistency;
+        private readonly IWriteConsistency _writeConsistency;
+        private const int MaxUpdateAttempts = 3;
+
+        // The default maximum-frame-size is 256 KiB with Artery.
+        // When using entity identifiers with 36 character strings (e.g. UUID.randomUUID).
+        // By splitting the elements over 5 keys we can support 10000 entities per shard.
+        // The Gossip message size of 5 ORSet with 2000 ids is around 200 KiB.
```

Comment:
These values are the same in .NET Desktop, slightly smaller in .NET Core since strings are UTF-8 instead of Unicode there. But either way, we're still in the clear.